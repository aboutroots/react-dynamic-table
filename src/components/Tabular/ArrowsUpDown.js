import React from "react";

import { FaAngleDown, FaAngleUp } from "react-icons/fa";

export class ArrowsUpDown extends React.Component {
  handleUp = () => {};

  handleDown = () => {};

  render() {
    return (
      <div className="ArrowsUpDown">
        <FaAngleUp onClick={this.handleUp} />
        <FaAngleDown onClick={this.handleDown} />
      </div>
    );
  }
}

export default ArrowsUpDown;
