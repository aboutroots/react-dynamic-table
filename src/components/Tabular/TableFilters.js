// @flow
import React from 'react';

import { connect } from 'react-redux';
import { setItemsOnPageFilter, setUserNameFilter } from '../../actions/filters';

type Props = {
  setItemsOnPageFilter: (itemsOnPage: number) => void,
  setUserNameFilter: (userNameFilter: string) => void,
  userNameFilter: string,
  itemsOnPage: number,
  users: Array<string>,
};

export class TableFilters extends React.Component<Props> {
  handleItemsOnPageChange = (e: SyntheticEvent<HTMLSelectElement>) => {
    const itemsOnPage = parseInt(e.currentTarget.value);
    this.props.setItemsOnPageFilter(itemsOnPage);
  };

  handleUserNameChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
    // TODO: handle autosugestions using props.users
    const userNameFilter = e.currentTarget.value;
    this.props.setUserNameFilter(userNameFilter);
  };

  render() {
    return (
      <div className="TableFilters">
        <div className="TableFilters__filter">
          <p className="TableFilters__title">Filter by user</p>
          <input
            type="text"
            placeholder="Start writing..."
            value={this.props.userNameFilter}
            onChange={this.handleUserNameChange}
            className="inpt"
          />
        </div>
        <div className="TableFilters__filter">
          <p className="TableFilters__title">Number of visible rows</p>
          <select
            value={this.props.itemsOnPage}
            onChange={this.handleItemsOnPageChange}
          >
            {[5, 10, 15, 20].map(o => (
              <option key={o}>{o}</option>
            ))}
          </select>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ filters, posts }) => ({
  itemsOnPage: filters.itemsOnPage,
  userNameFilter: filters.userNameFilter,
  users: posts.map(post => post.userName),
});

const mapDispatchToProps = dispatch => ({
  setUserNameFilter: value => dispatch(setUserNameFilter(value)),
  setItemsOnPageFilter: value => dispatch(setItemsOnPageFilter(value)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TableFilters);
