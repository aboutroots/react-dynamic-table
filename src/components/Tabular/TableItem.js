import * as React from "react";

export default function TableItem() {
  return (
    <div className="TableItem">
      <p className="TableItem__col" />
    </div>
  );
}
