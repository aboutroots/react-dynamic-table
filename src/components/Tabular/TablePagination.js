// @flow
import React from 'react';
import { connect } from 'react-redux';

import { setCurrentPage } from '../../actions/filters';
import { selectFiltered } from '../../selectors/posts';

type Props = {
  setCurrentPage: (page: number) => void,
  currentPage: number,
  pagesCount: number,
};

export class TablePagination extends React.Component<Props> {
  handleOnClick = (page: number) => {
    this.props.setCurrentPage(page);
  };
  render() {
    return (
      <div className="TablePagination">
        {[...Array(this.props.pagesCount).keys()].map(i => {
          let className = 'TablePagination__page';
          if (i === this.props.currentPage) {
            className += ' TablePagination__page--active';
          }
          return (
            <div
              className={className}
              key={i}
              onClick={() => this.handleOnClick(i)}
            >
              {i + 1}
            </div>
          );
        })}
      </div>
    );
  }
}
const mapStateToProps = state => {
  const activeUser = state.auth.activeUser;
  const filteredPosts = selectFiltered(state.posts, activeUser, state.filters);
  const pagesCount = Math.ceil(
    filteredPosts.length / state.filters.itemsOnPage
  );
  return {
    currentPage: state.filters.currentPage,
    pagesCount,
  };
};

const mapDispatchToProps = dispatch => ({
  setCurrentPage: page => dispatch(setCurrentPage(page)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TablePagination);
