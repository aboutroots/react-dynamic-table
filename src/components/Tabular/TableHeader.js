import React from 'react';

import ArrowsUpDown from './ArrowsUpDown';

// mapping code names to human readable names
const cols = {
  id: '#ID',
  username: 'USERNAME',
  postTitle: 'POST TITLE',
  views: 'VIEWS',
  likes: 'LIKES',
  createdAt: 'CREATED AT',
};

export default function TableHeader() {
  return (
    <div className="TableHeader">
      {Object.entries(cols).map(([k, v]) => (
        <div className="TableHeader__col" key={k}>
          <p>{v}</p>
          <ArrowsUpDown columnName={k} />
        </div>
      ))}
    </div>
  );
}
