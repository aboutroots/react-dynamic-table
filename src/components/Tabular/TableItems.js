import React from 'react';
import { connect } from 'react-redux';

import TableItem from './TableItem';
import { selectFiltered, selectPaginated } from '../../selectors/posts';

export function TableItems(props) {
  return (
    <div className="TableItems">
      {props.posts.map((post, idx) => (
        <TableItem {...post} key={idx} />
      ))}
    </div>
  );
}
const mapStateToProps = state => {
  const activeUser = state.auth.activeUser;
  const filtered = selectFiltered(state.posts, activeUser, state.filters);
  const paginated = selectPaginated(filtered, state.filters);
  return {
    posts: paginated,
  };
};

export default connect(mapStateToProps)(TableItems);
