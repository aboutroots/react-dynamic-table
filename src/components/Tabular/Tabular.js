import React, { Component } from 'react';

import TableFilters from './TableFilters';
import TableHeader from './TableHeader';
import TableItems from './TableItems';
import TablePagination from './TablePagination';

export class Tabular extends Component {
  render() {
    return (
      <div className="Tabular">
        <TableFilters />
        <TableHeader />
        <TableItems />
        <TablePagination />
      </div>
    );
  }
}

export default Tabular;
