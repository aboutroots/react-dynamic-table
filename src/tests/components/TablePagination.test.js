import React from 'react';
import { shallow } from 'enzyme';
import { TablePagination } from '../../components/Tabular/TablePagination';

let wrapper, setCurrentPage, currentPage, pagesCount;

beforeEach(() => {
  setCurrentPage = jest.fn();
  currentPage = 0;
  pagesCount = 5;
  wrapper = shallow(
    <TablePagination
      setCurrentPage={setCurrentPage}
      currentPage={currentPage}
      pagesCount={pagesCount}
    />
  );
});

test('should render TablePagination', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should set current page', () => {
  wrapper
    .find('p')
    .at(1)
    .simulate('click');
  expect(setCurrentPage).toHaveBeenLastCalledWith(1);
});
