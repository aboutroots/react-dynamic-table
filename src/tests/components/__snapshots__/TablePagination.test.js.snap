// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TablePagination 1`] = `
<div
  className="TablePagination"
>
  <p
    className="TablePagination__page TablePagination__page--active"
    key="0"
    onClick={[Function]}
  >
    1
  </p>
  <p
    className="TablePagination__page"
    key="1"
    onClick={[Function]}
  >
    2
  </p>
  <p
    className="TablePagination__page"
    key="2"
    onClick={[Function]}
  >
    3
  </p>
  <p
    className="TablePagination__page"
    key="3"
    onClick={[Function]}
  >
    4
  </p>
  <p
    className="TablePagination__page"
    key="4"
    onClick={[Function]}
  >
    5
  </p>
</div>
`;
